File: src/main/scala/com/lp/test/sideoutputs/SideOutPutTest.scala
package com.lp.test.sideoutputs

import java.util.Properties

import com.lp.test.sideoutputs.LateDataSideOut.CustomWatermarkExtractor
import com.lp.test.utils.ConfigUtils
import org.apache.flink.api.common.functions.RichMapFunction
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer
import org.apache.flink.util.{Collector, OutputTag}

/**
  * <p/> 
  * <li>Description: 侧输出分流</li>
  * flink提供了侧输出这个功能，侧输出的输出类型可以与主流不同，可以有多个侧输出(sideoutput)，每个侧输出不同的类型。
  *     1. 定义OutputTag
  *     2. 使用特定的函数ProcessFunction、CoProcessFunction、ProcessWindowFunction、ProcessAllWindowFunction
  * <li>@author: lipan@cechealth.cn</li> 
  * <li>Date: 2019-05-10 13:28</li> 
  */
object SideOutPutTest {

  val outputTag = new OutputTag[String]("side_output>5")

  def main(args: Array[String]): Unit = {

    val env = StreamExecutionEnvironment.getExecutionEnvironment

    //选择设置事件时间和处理事件
    env.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime)
    val kafkaConfig = ConfigUtils.apply("kv")

    val kafkaConsumer = new FlinkKafkaConsumer(kafkaConfig._1,
      new SimpleStringSchema(), //自定义反序列化器
      kafkaConfig._2)
      .setStartFromEarliest()

    import org.apache.flink.api.scala._
    val mainStream = env.addSource(kafkaConsumer)
      .map(new RichMapFunction[String, Integer] {
        override def map(value: String): Integer = {
          Integer.valueOf(value)
        }
      })
      //为流中的每个元素调用该函数，并可以生成零或多个输出流。
      .process(new ProcessFunction[Integer, Integer] {
      override def processElement(value: Integer, ctx: ProcessFunction[Integer, Integer]#Context, out: Collector[Integer]): Unit = {
        if (value <= 5)
          out.collect(value) //小于5的数据正常输出
        else
          ctx.output(outputTag, "sideout-" + String.valueOf(value)) //侧输出，可能有多个
      }
    })
    mainStream.print()
    //    mainStream.getSideOutput(outputTag).print()
    env.execute("SideOutPutTest")
  }

}

File: src/main/java/com/lp/java/demo/datastream/sideoutputs/SideOutputTest.java
package com.lp.java.demo.datastream.sideoutputs;

import com.lp.scala.demo.utils.ConfigUtils;
import org.apache.flink.api.common.functions.RichMapFunction;
import org.apache.flink.api.common.restartstrategy.RestartStrategies;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumerBase;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;

import java.util.Properties;


/**
 * <p/>
 * <li>title: 侧输出-分流</li>
 * <li>@author: li.pan</li>
 * <li>Date: 2020/1/12 2:52 下午</li>
 * <li>Version: V1.0</li>
 * <li>Description: https://zhuanlan.zhihu.com/p/54583067
 * </li>
 */
public class SideOutputTest {
    final static OutputTag<String> outputTag = new OutputTag<String>("side-output>5") {
    };

    public static void main(String[] args) throws Exception {
        // set up the streaming execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();


        /**
         * 设置重启策略/5次尝试/每次尝试间隔50s
         */
        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(5, 50000));

        // 选择设置时间
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

        scala.Tuple2<String, Properties> kafkaConfig = ConfigUtils.apply("string");
        /**
         * 从最新的offset开始消费消息
         * 设置自定义时间戳分配器和watermark发射器，也可以在后面的算子中设置
         */
        FlinkKafkaConsumerBase kafkaConsumer =
                new FlinkKafkaConsumer(kafkaConfig._1, new SimpleStringSchema(), kafkaConfig._2)
                        .setStartFromLatest();
        SingleOutputStreamOperator mainStream = env.addSource(kafkaConsumer)
                .map(new String2Integer())
                .process(new ProcessFunction<Integer, Integer>() {

                    @Override
                    public void processElement(Integer value, Context ctx, Collector<Integer> out) throws Exception {
                        // emit data to regular output
                        if (value <= 5)
                            out.collect(value);
                        if (value > 5)
                            // emit data to side output
                            ctx.output(outputTag, "sideout-" + String.valueOf(value));
                    }
                });
        mainStream.print();
        mainStream.getSideOutput(outputTag).print();

        env.execute(SideOutputTest.class.getCanonicalName());
    }

    private static class String2Integer extends RichMapFunction<String, Integer> {
        private static final long serialVersionUID = 1180234853172462378L;

        @Override
        public Integer map(String event) throws Exception {

            return Integer.valueOf(event);
        }

        @Override
        public void open(Configuration parameters) throws Exception {
        }
    }
}
